[Feature] Supports Views and Materialized Views for starrocks-sqlalchemy + Docs Enhancement

jaogoy opened the pull request:
## Why I'm doing:

starrocks-sqlalchemy need to support View and Materialized View for object declearation and doing comparison (schema migration) by using Alembic.

## What I'm doing:

**Description:**

This PR introduces initial support for SQLAlchemy `View` and
`MaterializedView` objects within the `sqlalchemy-starrocks` dialect,
enabling their definition in Python models and integration with Alembic
for schema migrations. Additionally, it provides extensive documentation
covering both usage and internal design, and refines existing table
definition documentation.

**Design Overview:**

- **Unified Table-Based Model**: `View` and `MaterializedView` are
implemented as specialized `Table` objects. Internally they are
completely treated as `Table` instances, with the real kind recorded in
`table.info['table_kind']` (`TABLE` / `VIEW` / `MATERIALIZED_VIEW`).
This means in all layers (SQLAlchemy metadata, reflection, compiler,
Alembic compare/render), they can be operated on using the standard
`Table` APIs, with additional semantics derived from `table_kind` and
StarRocks-specific options.
- **Reflection via `reflect_table`**: The reflection design is centered
around `Inspector.reflect_table()`. Regardless of whether the underlying
object in StarRocks is a base table, view, or materialized view,
reflection always returns a `Table` object and fills
`table.info['table_kind']`, `table.info['definition']` and
`table.dialect_options['starrocks']` accordingly. All subsequent
components (DDL compilation, Alembic comparators, renderers) consume
this unified `Table` representation.
- **Core Design Doc**: The detailed design and rationale for this model
are documented in `docs/design/view_and_mv.md`, which describes how the
schema, reflection, compiler, compare, and render layers cooperate
around this unified representation.

**Key Features & Enhancements:**

1. **SQLAlchemy `View` Support:**
  
    - Introduced `starrocks.sql.schema.View` object for defining database
    views in Python.
    - Full parameter support for `View` (name, definition, schema,
    comment, columns, security).
    - Reflection capabilities for existing views (names, definitions,
    security types, etc.).

2. **SQLAlchemy `MaterializedView` Support:**
    
    - Introduced `starrocks.sql.schema.MaterializedView` object for
    defining materialized views.
    - Support for `MaterializedView` parameters (name, definition,
    schema, properties for refresh/partition/distribution).
    - Reflection capabilities for existing materialized views (names,
    definitions, properties, etc.).

3. **Alembic Integration for Views & Materialized Views:**
    
    - Custom Alembic operations (`CreateViewOp`, `DropViewOp`,
    `AlterViewOp`, `CreateMaterializedViewOp`, `DropMaterializedViewOp`,
    `AlterMaterializedViewop`) for DDL generation.
    - Enhanced `StarRocksDDLCompiler` to translate SQLAlchemy `View` and
    `MaterializedView` objects into correct StarRocks DDL.
    - Improved `autogenerate` comparison logic
    (`starrocks.alembic.compare`) to detect differences in view/MV
    definitions and generate migration scripts.
    - Custom renderers (`starrocks.alembic.render`) to convert Alembic
    `Op` objects into executable Python code for migration scripts.
    - Guidance on `env.py` configuration (`render_item`,
    `include_object`) for proper autogeneration of Views and MVs.

4. **Enhanced Table Definition & Column Properties:**
    
    - Standardized `starrocks_` prefixed keyword arguments for both
    `Table` and `Column` objects (e.g., `starrocks_agg_type='SUM'`).
    - Introduced `starrocks_is_agg_key` for explicitly marking key
    columns in `AGGREGATE KEY` tables.
    - Added detailed examples for various table types (Primary Key,
    Aggregate Key, Duplicate Key) and complex column types (ARRAY, MAP,
    STRUCT).

5. **Comprehensive Documentation (`docs/`)**:
   - **Usage Guides (`docs/usage_guide/`)**:
      - `tables.md`: Fully revised with detailed explanations of
      table-level and column-level `starrocks_` attributes, including usage
      examples and ORM declarative style.
      - `views.md`: Detailed guide on defining and managing views with
      SQLAlchemy and Alembic.
      - `materialized_views.md`: Detailed guide on defining and managing
      materialized views with SQLAlchemy and Alembic.
   - **Design Documents (`docs/design/`)**:
      - `view_and_mv.md`: Core design document that describes the unified
      Table-based model for views/materialized views and how it integrates
      with reflection, compiler, Alembic compare, and render.
      - `inspect.md`: Details the introspection strategy, including
      querying `information_schema` views (such as `tables`, `tables_conf`,
      `views`, `materialized_views`) and parsing `SHOW CREATE TABLE` output
      into structured objects and reflecting column aggregate types.
      - `compile.md`: Explains the DDL compilation process for
      StarRocks-specific constructs.
      - `compare.md`: Provides a deep dive into Alembic's comparison
      dispatch mechanism, including a flowchart and how the
      `sqlalchemy-starrocks` dialect customizes it.
      - `render.md`: Describes how migration operations are rendered into
      Python code.

**Important Considerations & Known Limitations (for this PR):**

- **Materialized View `ALTER` Operations**: While `CREATE` and `DROP`
are supported, comprehensive `ALTER MATERIALIZED VIEW` operations (e.g.,
for `PARTITION BY`, `DISTRIBUTED BY`, `STATUS`) are not supported by
StarRocks or still under development and not fully covered by
autogeneration in this PR.
- **SQLAlchemy Primary Key Requirement**: Due to SQLAlchemy's internal
validation, columns designated as StarRocks primary keys (e.g., via
`starrocks_primary_key`) must _also_ be explicitly marked as
`primary_key=True` in their `Column(...)` definition to prevent errors,
even if this SQLAlchemy-level declaration is not directly used by the
dialect for DDL generation.
- **`SECURITY` Type for Views**: The `DEFINER` security type is not yet
implemented in StarRocks (v3.5) and is noted in `views.md`.
- **`ALTER TABLE` Concurrency**: Long-running `ALTER TABLE` operations
(e.g., `MODIFY COLUMN`) in StarRocks can block subsequent schema changes
on the same table. Users should review generated migration scripts and
manually split operations if necessary (documented in `tables.md`).
- **Complex Type Column Modifications**: StarRocks has strict
limitations on altering columns with complex types (`ARRAY`, `MAP`,
`STRUCT`). Autogeneration may produce invalid DDL for such changes,
requiring manual intervention (documented in `tables.md`).

**Testing:**

- Unit and integration tests have been extended to cover the new `View`
and `MaterializedView` functionality, as well as the enhanced table and
column property handling..

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Implements first-class View and MaterializedView support (definition, reflection, DDL, Alembic autogenerate/render) with extensive docs and tests, plus related API refinements.
> 
> - **SQLAlchemy Dialect**:
>   - Add `View` and `MaterializedView` (as `Table`-based) with unified `table.info['table_kind']` and `info['definition']`.
>   - Reflection reworked via `Inspector.reflect_table()`; parses SECURITY, REFRESH, PARTITION/DISTRIBUTION/ORDER/PROPERTIES.
>   - DDL compiler dispatches `CREATE/DROP/ALTER` for View/MV; supports table-post clauses for MV.
>   - Standardize lowercase `starrocks_` options (e.g., `starrocks_partition_by`, `starrocks_properties`).
>   - Fix/adjust autoincrement alter emission, type/SQL normalization, defaults handling.
> - **Alembic Integration**:
>   - New ops: `Create/Drop/AlterViewOp`, `Create/Drop/AlterMaterializedViewOp`.
>   - Compare hooks for View/MV; `AlterMaterializedViewOp` handles refresh/properties; immutable changes raise.
>   - Renderers for View/MV ops; add `include_object_for_view_mv` and `combine_include_object` filters.
>   - Env guidance: `render_column_type`, `include_object`, multi-schema `include_name`.
> - **Docs**:
>   - New/expanded guides: Alembic, Views, Materialized Views, Tables; design docs (view/mv, compile, inspect, compare, render); test coverage summaries.
>   - README and examples updated (connection strings, usage).
> - **Tests**:
>   - Extensive unit/integration/system tests for View/MV reflection, compile, render, autogenerate, lifecycle; table/type/reflection updates; CI fixtures adjusted.
> - **Misc**:
>   - Bump version to `1.3.1`; expose `starrocks.schema`; normalize utilities; minor API consistency fixes.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6ca2e6f3e4ccc8fa3c5111d6072ad4a42ea5aea7. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jaogoy committed:
Feat: Initial Support for SQLAlchemy Views and Materialized Views + Docs Enhancement

**Description:**

This PR introduces initial support for SQLAlchemy `View` and
`MaterializedView` objects within the `sqlalchemy-starrocks` dialect,
enabling their definition in Python models and integration with Alembic
for schema migrations. Additionally, it provides extensive documentation
covering both usage and internal design, and refines existing table
definition documentation.

**Design Overview:**

- **Unified Table-Based Model**: `View` and `MaterializedView` are
implemented as specialized `Table` objects. Internally they are
completely treated as `Table` instances, with the real kind recorded in
`table.info['table_kind']` (`TABLE` / `VIEW` / `MATERIALIZED_VIEW`).
This means in all layers (SQLAlchemy metadata, reflection, compiler,
Alembic compare/render), they can be operated on using the standard
`Table` APIs, with additional semantics derived from `table_kind` and
StarRocks-specific options.
- **Reflection via `reflect_table`**: The reflection design is centered
around `Inspector.reflect_table()`. Regardless of whether the underlying
object in StarRocks is a base table, view, or materialized view,
reflection always returns a `Table` object and fills
`table.info['table_kind']`, `table.info['definition']` and
`table.dialect_options['starrocks']` accordingly. All subsequent
components (DDL compilation, Alembic comparators, renderers) consume
this unified `Table` representation.
- **Core Design Doc**: The detailed design and rationale for this model
are documented in `docs/design/view_and_mv.md`, which describes how the
schema, reflection, compiler, compare, and render layers cooperate
around this unified representation.

**Key Features & Enhancements:**

1. **SQLAlchemy `View` Support:**

- Introduced `starrocks.sql.schema.View` object for defining database
views in Python.
- Full parameter support for `View` (name, definition, schema,
comment, columns, security).
- Reflection capabilities for existing views (names, definitions,
security types, etc.).

2. **SQLAlchemy `MaterializedView` Support:**

- Introduced `starrocks.sql.schema.MaterializedView` object for
defining materialized views.
- Support for `MaterializedView` parameters (name, definition,
schema, properties for refresh/partition/distribution).
- Reflection capabilities for existing materialized views (names,
definitions, properties, etc.).

3. **Alembic Integration for Views & Materialized Views:**

- Custom Alembic operations (`CreateViewOp`, `DropViewOp`,
`AlterViewOp`, `CreateMaterializedViewOp`, `DropMaterializedViewOp`,
`AlterMaterializedViewop`) for DDL generation.
- Enhanced `StarRocksDDLCompiler` to translate SQLAlchemy `View` and
`MaterializedView` objects into correct StarRocks DDL.
- Improved `autogenerate` comparison logic
(`starrocks.alembic.compare`) to detect differences in view/MV
definitions and generate migration scripts.
- Custom renderers (`starrocks.alembic.render`) to convert Alembic
`Op` objects into executable Python code for migration scripts.
- Guidance on `env.py` configuration (`render_item`,
`include_object`) for proper autogeneration of Views and MVs.

4. **Enhanced Table Definition & Column Properties:**

- Standardized `starrocks_` prefixed keyword arguments for both
`Table` and `Column` objects (e.g., `starrocks_agg_type='SUM'`).
- Introduced `starrocks_is_agg_key` for explicitly marking key
columns in `AGGREGATE KEY` tables.
- Added detailed examples for various table types (Primary Key,
Aggregate Key, Duplicate Key) and complex column types (ARRAY, MAP,
STRUCT).

5. **Comprehensive Documentation (`docs/`)**:
   - **Usage Guides (`docs/usage_guide/`)**:
- `tables.md`: Fully revised with detailed explanations of
table-level and column-level `starrocks_` attributes, including usage
examples and ORM declarative style.
- `views.md`: Detailed guide on defining and managing views with
SQLAlchemy and Alembic.
- `materialized_views.md`: Detailed guide on defining and managing
materialized views with SQLAlchemy and Alembic.
   - **Design Documents (`docs/design/`)**:
- `view_and_mv.md`: Core design document that describes the unified
Table-based model for views/materialized views and how it integrates
with reflection, compiler, Alembic compare, and render.
- `inspect.md`: Details the introspection strategy, including
querying `information_schema` views (such as `tables`, `tables_conf`,
`views`, `materialized_views`) and parsing `SHOW CREATE TABLE` output
into structured objects and reflecting column aggregate types.
- `compile.md`: Explains the DDL compilation process for
StarRocks-specific constructs.
- `compare.md`: Provides a deep dive into Alembic's comparison
dispatch mechanism, including a flowchart and how the
`sqlalchemy-starrocks` dialect customizes it.
- `render.md`: Describes how migration operations are rendered into
Python code.

**Important Considerations & Known Limitations (for this PR):**

- **Materialized View `ALTER` Operations**: While `CREATE` and `DROP`
are supported, comprehensive `ALTER MATERIALIZED VIEW` operations (e.g.,
for `PARTITION BY`, `DISTRIBUTED BY`, `STATUS`) are not supported by
StarRocks or still under development and not fully covered by
autogeneration in this PR.
- **SQLAlchemy Primary Key Requirement**: Due to SQLAlchemy's internal
validation, columns designated as StarRocks primary keys (e.g., via
`starrocks_primary_key`) must _also_ be explicitly marked as
`primary_key=True` in their `Column(...)` definition to prevent errors,
even if this SQLAlchemy-level declaration is not directly used by the
dialect for DDL generation.
- **`SECURITY` Type for Views**: The `DEFINER` security type is not yet
implemented in StarRocks (v3.5) and is noted in `views.md`.
- **`ALTER TABLE` Concurrency**: Long-running `ALTER TABLE` operations
(e.g., `MODIFY COLUMN`) in StarRocks can block subsequent schema changes
on the same table. Users should review generated migration scripts and
manually split operations if necessary (documented in `tables.md`).
- **Complex Type Column Modifications**: StarRocks has strict
limitations on altering columns with complex types (`ARRAY`, `MAP`,
`STRUCT`). Autogeneration may produce invalid DDL for such changes,
requiring manual intervention (documented in `tables.md`).

**Testing:**

- Unit and integration tests have been extended to cover the new `View`
and `MaterializedView` functionality, as well as the enhanced table and
column property handling..

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
refine docs

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
fix for auto_increment

1. because the previous PR supports auto_increment
2. some small bugs: bindparams for information_schema.xxx, cases for
auto increment.

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
refine code according to robot review

Signed-off-by: jaogoy <jaogoy@gmail.com>

rad-pat commented on contrib/starrocks-python-client/docs/test/test_coverage.md:
```diff
@@ -8,82 +8,212 @@ This suite verifies that Python schema definitions are correctly compiled into S
+2. **Integration Tests** (`test/integration/`): Test interactions with a real StarRocks database (reflection, autogenerate comparison).
+3. **System Tests** (`test/system/`): End-to-end tests simulating the complete user workflow from schema definition through migration execution and rollback.
+
+The **system tests** are particularly important as they validate the entire Alembic lifecycle that users will experience in production, ensuring that:
```
Probably should extend from the Alembic Test Suite provided by alembic - which integrates with the SQLAlchemy test suite, then all tests can be ran together?

jaogoy committed:
fix test bugs

Signed-off-by: jaogoy <jaogoy@gmail.com>

rad-pat reviewed (changes requested):
Reverting your incorrect change to autoincrement behaviour fixes the test_suite to just one failure. If you are using `primary_key` flag in other tests, you need to also mark as `autoincrement=False` if it is a single integer column.

jaogoy commented:
@zhaohehuhu If you have time, I will be very glad to see your review and comments.

jaogoy commented:
> Ok, @jaogoy, I will look to find the test override - I already made it somewhere for previous PR - if not then I can quickly make a commit that you can cherry pick to your branch - if this is OK with you? For ref, here is the EnumTest in SQLAlchemy code - https://github.com/sqlalchemy/sqlalchemy/blob/6ee4d884f04aa6bad21ad280a02441262863394c/lib/sqlalchemy/testing/suite/test_types.py#L2034

Ooh, It's my carelessness. The source code is old, and the class is not imported.
Now it works for the `EnumTest` and other similar.

@rad-pat Thanks a lot.

jaogoy committed:
fix slqlalchemy spec test

1. It need to call reflect_table first for unreflectable tables.
2. Manually declare tables with autoincrement=True for test cases,
because for StarRocks, autoincrement is False by default.

Signed-off-by: jaogoy <jaogoy@gmail.com>

rad-pat reviewed (approved):
All good, just a comment

alvin-celerdata commented:
@cursor review
